### PR TITLE
(artifact-parser): Add the Version header parser

### PR DESCRIPTION
**Pull request by oleorhagen**
This adds a simple parser for the Mender-Artifact v3 version header.

As well as a small utility ByteWriter class to io.

I do have to admit that I feel this turned out needlessly complicated, so all open for advice on how to improve this!

### CODE REVIEW

File: artifact/version/version.cpp
Comment (kacf) [inline artifact/version/version.cpp:70]: I'm somewhat confused by this code. Is there three layers of ExpectedJson inside each other? Line 64 looks like a no-op, you are just assigning a payload value to an ExpectedJson, and since it is not an error, it will always construct an ExpectedJson with a value, and the next `if` will always evaluate false.

But feel free to enlighten me if I'm misunderstanding this.

File: artifact/version/version.cpp
Comment (kacf) [inline artifact/version/version.cpp:74]: The JSON is valid though, so should this be `FormatError` instead? It's JSON, but a format we don't understand.

File: artifact/version/version.cpp
Comment (kacf) [inline artifact/version/version.cpp:81]: Same here, maybe `FormatError`.

File: artifact/version/version.cpp
Comment (kacf) [inline artifact/version/version.cpp:85]: Trailing underscores are for member variables.

File: artifact/version/version.hpp
Comment (kacf) [inline artifact/version/version.hpp:11]: 👍

But should the files be inside `v3` as well?

File: artifact/version/version.cpp
Comment (kacf) [inline artifact/version/version.cpp:104]: Suggestion: You could shorten this by doing:

```c++
auto format = expected_obj.Get("format");
if (!format || !format.value().GetString()) {
	return MakeError(
		ParseError,
		"Failed to unmarshal the version from the version JSON data. 'format' field not found or not a string");
}
auto format_unwrapped = format.value().GetString();
```

And similar with `version` above. Not saying it's better though, verbosity is always good. Just a suggestion.

File: artifact/version/version.cpp
Comment (kacf) [inline artifact/version/version.cpp:115]: > Failed to unmarshal the version from the version JSON data

You've used this message in nearly every error so far. Maybe you should make it part of the error code string instead, and use only the more specific strings in this function?

File: artifact/version/version.hpp
Comment (kacf) [inline artifact/version/version.hpp:25]: I have a question about this, but it's an abstract question, hard to put into words here. Ping me when you have time.

File: artifact/version/version.cpp
Comment (oleorhagen) [reply artifact/version/version.cpp:70]: I don't even know. I think this is horrible...

However, I have not come up with a cleaner way...

Comment (oleorhagen) [reply artifact/version/version.cpp:70]: Alright, cleaning it up a little...

File: artifact/v3/version/version.cpp
Comment (vpodzime) [inline artifact/v3/version/version.cpp:78]: Might as well just return `expected_json.error()` here.

File: artifact/v3/version/version.cpp
Comment (vpodzime) [inline artifact/v3/version/version.cpp:75]: We copy the data into a buffer only to create a string from it here and throw the buffer away. It would be needed if `StringFromByteVector` took a reference and potentially modified the given buffer, but it is (I believe) a `const` reference and so we can safely pass in a buffer without creating a copy of it.

File: artifact/v3/version/version.cpp
Comment (vpodzime) [inline artifact/v3/version/version.cpp:70]: No a real issue here, I believe, but what if the whole JSON doesn't fit into the buffer?

File: artifact/v3/version/version.cpp
Comment (vpodzime) [inline artifact/v3/version/version.cpp:94]: Out of  scope of this PR, but `ExpectedInt` should be in the `expected` namespace/header, I think.

Comment (vpodzime) [reply artifact/v3/version/version.cpp:94]: Please read this line (maybe out loud) and I'm pretty sure you will be able to come up with better variable names.

File: artifact/v3/version/version.cpp
Comment (vpodzime) [inline artifact/v3/version/version.cpp:101]: Failing to get the version integer should also be an error, right?

File: artifact/v3/version/version.cpp
Comment (vpodzime) [inline artifact/v3/version/version.cpp:106]: `if (format_value)`?

File: artifact/v3/version/version.cpp
Comment (vpodzime) [inline artifact/v3/version/version.cpp:114]: Again, `!value_format` is an error too, right?

File: artifact/v3/version/version.cpp
Comment (vpodzime) [inline artifact/v3/version/version.cpp:119]: It feels a bit strange to have a function called `Parse()` that doesn't actually return any parsed data, but not a big deal.

File: artifact/v3/version/version.hpp
Comment (vpodzime) [inline artifact/v3/version/version.hpp:55]: Similar here, a bit weird to have a type for something that's actually just a hard-coded constant pair `3-"mender"`.

File: artifact/v3/version/version_test.cpp
Comment (vpodzime) [inline artifact/v3/version/version_test.cpp:20]: `sstream`?

File: artifact/v3/version/version.cpp
Comment (kacf) [inline artifact/v3/version/version.cpp:97]: Come to think of it, maybe this should be a constant.

File: artifact/v3/version/version.cpp
Comment (kacf) [inline artifact/v3/version/version.cpp:95]: This logic is wrong now. If the `version` field is not an int, the code continues, but it should fail.

File: artifact/v3/version/version.cpp
Comment (kacf) [reply artifact/v3/version/version.cpp:106]: Yes, plus code continues on error here as well.

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp:26]: Hmm... Why not?

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp:75]: The absolute best would be if `json` took a `Reader`, and I wouldn't have to do any of this.

Since ur intimately familiar with this. Would it be a lot of work to add you think?

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp:70]: Then we've received some mumbo-jumbo which we don't support anyways :smile:

**Review by kacf — CHANGES_REQUESTED** [review]

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp:94]: hahaha. I think so too

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp:119]: Well, it does, right. The parsed values match exactly this. I just wanted to highlight it here. But if it's odd I can change it.

File: artifact/v3/version/version.hpp
Comment (oleorhagen) [reply artifact/v3/version/version.hpp:55]: Maybe, but it does match the structure of what is parsed to a code-template of sorts. So at least it's symmetric with how we handle the rest of the parsing.

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp:95]: Oh

My

God....

File: artifact/v3/version/version.cpp
Comment (vpodzime) [reply artifact/v3/version/version.cpp:75]: Yeah, I think I can do that. The JSON stuff we have predates our IO stuff so it wasn't possible back then. The only complication is that to do it right, we should have a UTF8Reader because bytes shouldn't be passed into the JSON parser that expects characters.

Comment (oleorhagen) [reply artifact/v3/version/version.cpp:75]: Great :tada: :tada:

File: common/io.hpp
Comment (oleorhagen) [reply common/io.hpp:114]: I should really add a static check for this in my editor or smthn...

File: common/io.hpp
Comment (oleorhagen) [reply common/io.hpp:125]: We can unroll the loop? :p 

But really, I'll wrap it one of the std::copy derivatives

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp:78]: Can't directly actually, since they are different types...

But I can add a conversion function?

**Comment by oleorhagen** [comment]
Rebased on top of: https://github.com/mendersoftware/mender/pull/1173

File: artifact/v3/version/version.cpp
Comment (vpodzime) [inline artifact/v3/version/version.cpp:83]: Is this what our formatter produces? :scream:

Comment (vpodzime) [inline artifact/v3/version/version.cpp:83]: I'd drop the `value_or()` and got the `ExpectedInt()` from the chain because `value_or()` discards the error.

File: artifact/v3/version/version.cpp
Comment (vpodzime) [inline artifact/v3/version/version.cpp:88]: Should use the exact error (message) our JSON handling has the ability to provide.

File: artifact/v3/version/version.cpp
Comment (vpodzime) [inline artifact/v3/version/version.cpp:93]: Same here.

File: artifact/v3/version/version.cpp
Comment (vpodzime) [inline artifact/v3/version/version.cpp:101]: Should use the constants.

**Review by vpodzime — COMMENTED** [review]
+Still not a big fan of using the expected library without even trying to expand our own code, but looks like you guys are decided.

File: artifact/v3/version/CMakeLists.txt
Comment (oleorhagen) [reply artifact/v3/version/CMakeLists.txt:12]: Yeh, it's good you pointed this out, I forgot to mention it.

So I _did not_ make a module out of this, since it is so small, in which case, the test needs this to link.

However maybe this was a bad idea. I really don't know. 

Since it caught you by surprise, maybe I should make a separate library out of it, it just felt a bit extensive.

I'll let you decide :pray:

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp:23]: Yes, true. This is actually a really good rule!

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp:83]: Yes :see_no_evil: Not the prettiest on the :eyes:

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp:101]: Yeh, let's. I left it there to highlight it, but it's really just duplicating...

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp:64]: Yes, but like we discussed previously, the very best would be `json::From(io::Reader&)`

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp:68]: At least from what I'm used to, `Copy` is nice ,because it guarantees to read to the end. A single `Read` call can theoretically do short reads, and may need to be called again.

Comment (vpodzime) [reply artifact/v3/version/version.cpp:68]: But the data is shuffled around (into an internal buffer in `Copy` and then into the `ByteWriter` above) and you can easily get an error if `buf` above is not big enough. Like I said above, to me it would make much more sense to have a function/class/... to read everything into a buffer and then use that buffer.

File: artifact/v3/version/version.cpp
Comment (vpodzime) [reply artifact/v3/version/version.cpp:64]: That will require that function to read a file into a buffer anyway. Or a stream interface to our `Reader` because C++ already has IO utilities -- the streams -- an things are setup to work with them not our Go-like implementation.

File: artifact/v3/version/version.cpp
Comment (vpodzime) [reply artifact/v3/version/version.cpp:83]: `value_or()` is still there.

**Review by vpodzime — COMMENTED** [review]
DId you forget to push some commits? Many of the conversations are marked as resolved without the changes in the code being done.

**Comment by oleorhagen** [comment]
> DId you forget to push some commits? Many of the conversations are marked as resolved without the changes in the code being done.

Yes, just fixed locally, not pushed yet. I didn't know I had re-requested a review here, I'm sorry.
Fixed and pushed now @vpodzime :smile_cat:

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp:65]: This is true, it is just a question about the size of the buffer. Like buffering the whole payload, or doing it in chunks.

I actually had a conversation with @kacf about it this morning, where I suggested to move to the stdlib interface, and his argument is that the `io-streams` add too much code bloat for us, as they are all templated.

I'm not quite sure I fully agree, but I do understand the argument, and it is not our biggest thorn in our sides imo.

Therefore for now, going with this intermediary buffering is the way it seems.

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp]: Ur just up too early in the morning :p :sunrise_over_mountains:  :rooster:

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp:69]: This is true in the general case. But the advantage here is that we control the format, and the `version` header is never going to be bigger than what is here available.

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp]: Same as above. You need to sleep in more :laughing:

**Comment by oleorhagen** [comment]
I will not merge this until the `byte-writer` PR is merged!

**Review by kacf — APPROVED** [review]

File: artifact/v3/version/version.cpp
Comment (oleorhagen) [reply artifact/v3/version/version.cpp:69]: > But the data is shuffled around (into an internal buffer in Copy and then into the ByteWriter above) and you can easily get an error if buf above is not big enough. Like I said above, to me it would make much more sense to have a function/class/... to read everything into a buffer and then use that buffer.

Isn't this what I am doing with the `byte-buffer` above already?

**Comment by oleorhagen** [comment]
> I will not merge this until the `byte-writer` PR is merged!

Merged!

File: artifact/v3/version/version.cpp
Comment (vpodzime) [reply artifact/v3/version/version.cpp:69]: Assuming it's at most 1024 bytes. But let's not ride a dead horse here anymore. I'll try to come up with something for this myself and then we can have a discussion there.

**Review by vpodzime — APPROVED** [review]
Just a "nitpick of nitpicks". ACK!

File: artifact/v3/version/version.cpp
Comment (vpodzime) [reply artifact/v3/version/version.cpp:69]: > So there are 3 copies of the same thing, all with read-only access.

It's 4 actually. `Copy` has its own internal buffer too.

**Pull request merged**
